sampler.go

```diff
@@ -174,6 +181,177 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
+// rateLimitingSampler. The probabilisticSampler is given higher priority when tags are emitted,
+// ie. if IsSampled() for both samplers return true, the tags for probabilisticSampler will be used.
+func NewOperationSampler(operation string, maxTracesPerSecond, samplingRate float64) (Sampler, error) {
+	probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
+	if err != nil {
+		return nil, err
+	}
+	lowerBoundSampler, err := NewLowerBoundSampler(maxTracesPerSecond, samplingRate)
+	if err != nil {
+		return nil, err
+	}
+	return &operationSampler{
+		probabilisticSampler: probabilisticSampler,
+		lowerBoundSampler:    lowerBoundSampler,
+		operation:            operation,
+	}, nil
+}
+
+func (s *operationSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	if sampled, tags := s.probabilisticSampler.IsSampled(id, operation); sampled {
+		return true, tags
+	}
+	return s.lowerBoundSampler.IsSampled(id, operation)
+}
+
+func (s *operationSampler) Close() {
+	s.probabilisticSampler.Close()
+	s.lowerBoundSampler.Close()
+}
+
+func (s *operationSampler) Equal(other Sampler) bool {
+	if o, ok := other.(*operationSampler); ok {
+		return s.operation == o.operation && s.probabilisticSampler.Equal(o.probabilisticSampler) &&
+			s.lowerBoundSampler.Equal(o.lowerBoundSampler)
+	}
+	return false
+}
+
+// -----------------------
+
+type adaptiveSampler struct {
+	samplers map[string]Sampler
+
+	// TODO expand to keep track of newly seen operations
+	// TODO add a cap on the number of operations
+	// TODO need default rate limit and probabilistic rates for first time operations
+}
+
+// NewAdaptiveSampler adaptiveSampler is a delegating sampler that applies both probabilisticSampler and
+// rateLimitingSampler via the operationSampler. This sampler keeps track of all operations and delegates
+// calls to the respective operationSampler.
+func NewAdaptiveSampler(maxTracesPerSecond float64, samplingRates map[string]float64) (Sampler, error) {
+	samplers := make(map[string]Sampler)
+	for operation, samplingRate := range samplingRates {
+		probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
+		if err != nil {
+			return nil, err
+		}
+		lowerBoundSampler, err := NewLowerBoundSampler(maxTracesPerSecond, samplingRate)
+		if err != nil {
+			return nil, err
+		}
+		samplers[operation] = &operationSampler{
+			probabilisticSampler: probabilisticSampler,
+			lowerBoundSampler:    lowerBoundSampler,
+			operation:            operation,
+		}
+	}
+	return &adaptiveSampler{
+		samplers: samplers,
+	}, nil
+}
+
+func (s *adaptiveSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	sampler, ok := s.samplers[operation]
+	if !ok {
+		// TODO handle first time operation
+		return false, nil
+	}
+	return sampler.IsSampled(id, operation)
+}
+
+func (s *adaptiveSampler) Close() {
+	for _, sampler := range s.samplers {
+		sampler.Close()
+	}
+}
+
+func (s *adaptiveSampler) Equal(other Sampler) bool {
```

Comment:
the equals method is used to ensure RemoteSampler does not replace the existing sampler with a new instance if the sampler is the same. The primary reason was for rate limiting samplers that have internal state which we don't want to keep re-setting every minute.

That functionality is now pretty much mute in the RemoteSampler, because it always contain the adaptive sampler. On the other hand, we don't want to be re-creating the adaptive sampler and all its maps every minute, so it needs an intelligent way of taking updates for the sampling strategy and not replacing existing samplers if they haven't changed.

We may need to rethink the role of remote sampler vs. adaptive sampler. Perhaps there should be only one.